### PR TITLE
Refactor ADTask and Related Components

### PR DESCRIPTION
**Pull request by kaituo**
### Description
This PR includes several key refactoring changes:
- Extracts common code from ADTask into TimeSeriesTask, creating ForecastTask for forecasting-specific logic.
- Consolidates common code from ADTaskType into TaskType and introduces ForecastTaskType for forecasting-related purposes.
- Renames ADTaskState to TaskState for consistent code reuse.
- Renames the method getId in ADTask to getConfigId to differentiate it from other IDs like task id.

**Note:** The code refactored in this commit is subject to change when working on historical forecasting in the future.

Testing done:
1. Added unit tests for the new code to ensure functionality.
2. Executed a successful Gradle build.

### Issues Resolved
https://github.com/opensearch-project/anomaly-detection/issues/866

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/forecast/model/ForecastTaskType.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        );
+    public static List<ForecastTaskType> REALTIME_TASK_TYPES = ImmutableList
+        .of(ForecastTaskType.FORECAST_REALTIME_SINGLE_STREAM, ForecastTaskType.FORECAST_REALTIME_HC_FORECASTER);
+    public static List<ForecastTaskType> ALL_FORECAST_TASK_TYPES = ImmutableList
+        .of(
+            ForecastTaskType.FORECAST_REALTIME_SINGLE_STREAM,
+            ForecastTaskType.FORECAST_REALTIME_HC_FORECASTER,
+            ForecastTaskType.FORECAST_HISTORICAL_SINGLE_STREAM,
+            ForecastTaskType.FORECAST_HISTORICAL_HC_FORECASTER,
+            ForecastTaskType.FORECAST_HISTORICAL_HC_ENTITY
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply src/main/java/org/opensearch/forecast/model/ForecastTaskType.java:45]: I see the comments above -- is the goal to set up single entity HISTORICAL_HC_ENTITY ? (I was then thrown off by the "HC" prefix)

File: src/test/java/org/opensearch/timeseries/TestHelpers.java [hunk 2]
```diff
@@ -1427,6 +1420,20 @@ public static ADTask randomAdTask(
...
 
+    public static Entity randomEntity(Config config) {
+        Entity entity = null;
+        if (config != null) {
+            if (config.hasMultipleCategories()) {
+                Map<String, Object> attrMap = new HashMap<>();
+                config.getCategoryFields().stream().forEach(f -> attrMap.put(f, randomAlphaOfLength(5)));
+                entity = Entity.createEntityByReordering(attrMap);
+            } else if (config.isHighCardinality()) {
+                entity = Entity.createEntityByReordering(ImmutableMap.of(config.getCategoryFields().get(0), randomAlphaOfLength(5)));
```
Comment (sudiptoguha) [inline src/test/java/org/opensearch/timeseries/TestHelpers.java:1431]: Have a comment describing this?

**Review by sudiptoguha — DISMISSED** [review]
Looks good - a few q's.

File: src/main/java/org/opensearch/ad/model/ADTask.java [hunk 3]
```diff
@@ -185,175 +140,34 @@ public static Builder builder() {
         return new Builder();
     }
 
-    public boolean isHistoricalTask() {
-        return taskType.startsWith(HISTORICAL_TASK_PREFIX);
-    }
-
+    @Override
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/ad/model/ADTask.java:143]: We create a new task for each entity specific analysis in historical run. isEntityTask tells whether the task is on the entity time series level for historical run.

File: src/main/java/org/opensearch/forecast/model/ForecastTaskType.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/forecast/model/ForecastTaskType.java:45]: Is my following comment ok?

In real-time forecasting, we monitor states at the forecaster level, resulting in two distinct task types: one for single-stream forecasting and another for high cardinality (HC). In the historical setting, state tracking is more nuanced, encompassing both entity and forecaster levels. This leads to three specific task types: a forecaster-level task dedicated to single-stream forecasting, and two tasks for HC, one at the forecaster level and another at the entity level.

File: src/test/java/org/opensearch/timeseries/TestHelpers.java (hunk 2 above)
Comment (sudiptoguha) [reply src/test/java/org/opensearch/timeseries/TestHelpers.java:1431]: Thanks!

File: src/main/java/org/opensearch/forecast/model/ForecastTaskType.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/forecast/model/ForecastTaskType.java:45]: Changed to

```
/**
 * The ForecastTaskType enum defines different task types for forecasting, categorized into real-time and historical settings.
 * In real-time forecasting, we monitor states at the forecaster level, resulting in two distinct task types: one for
 * single-stream forecasting and another for high cardinality (HC). In the historical setting, state tracking is more nuanced,
 * encompassing both entity and forecaster levels. This leads to three specific task types: a forecaster-level task dedicated
 * to single-stream forecasting, and two tasks for HC, one at the forecaster level and another at the entity level.
 *
 * Real-time forecasting:
 * - FORECAST_REALTIME_SINGLE_STREAM: Represents a task type for single-stream forecasting. Ideal for scenarios where a single
 *  time series is processed in real-time.
 * - FORECAST_REALTIME_HC_FORECASTER: Represents a task type for high cardinality (HC) forecasting. Used when dealing with a
 *  large number of distinct entities in real-time.
 *
 * Historical forecasting:
 * - FORECAST_HISTORICAL_SINGLE_STREAM: Represents a forecaster-level task for single-stream historical forecasting.
 * Suitable for analyzing a single time series in a sequential manner.
 * - FORECAST_HISTORICAL_HC_FORECASTER: A forecaster-level task to track overall state, initialization progress, errors, etc.,
 * for HC forecasting. Central to managing multiple historical time series with high cardinality.
 * - FORECAST_HISTORICAL_HC_ENTITY: An entity-level task to track the state, initialization progress, errors, etc., of a
 * specific entity within HC historical forecasting. Allows for fine-grained information recording at the entity level.
 *
 */
```

[the remark this answers is not in the document]
Comment (sudiptoguha) [reply src/main/java/org/opensearch/forecast/model/ForecastTaskType.java:45]: Nice.

**Comment by kaituo** [comment]
@sudiptoguha added comments and addressed compiler errors due to https://github.com/opensearch-project/OpenSearch/pull/9006 

org.opensearch.common.component.LifecycleListener => org.opensearch.common.lifecycle.LifecycleListener
org.opensearch.common.unit.ByteSizeValue => org.opensearch.core.common.unit.ByteSizeValue
org.opensearch.indices.breaker.NoneCircuitBreakerService => org.opensearch.core.indices.breaker.NoneCircuitBreakerService
XContentFactory.contentBuilder => MediaTypeRegistry.contentBuilder

**Review by sudiptoguha — DISMISSED** [review]

File: src/main/java/org/opensearch/timeseries/model/TaskState.java [hunk 4]
```diff
@@ -9,47 +9,47 @@
...
- *     After task created, coordinate node will gather all eligible node’s state and dispatch
+ *     AD: After task created, coordinate node will gather all eligible node’s state and dispatch
  *     task to the worker node with lowest load. When the worker node receives the request,
  *     it will set the task state as INIT immediately, then start to run cold start to train
  *     RCF model. We will track the initialization progress in task.
  *     Init_Progress=ModelUpdates/MinSampleSize
  *
  * <li><code>RUNNING</code>:
- *     If RCF model gets enough data points and passed training, it will start to detect data
+ *     AD: If RCF model gets enough data points and passed training, it will start to detect data
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/timeseries/model/TaskState.java:33]: Are these all `AD` for now until you add the forecasting use case?

Comment (kaituo) [reply src/main/java/org/opensearch/timeseries/model/TaskState.java:33]: yes, forecast state are different from AD.

**Comment by kaituo** [comment]
> Have you done any testing with frontend for this branch?

I haven't. The frontend main branch CI will test backend main branch. If there is anything broken, I'll fix it. Also, since we won't release this branch until forecasting frontend is finished, I think we can leave the testing later.

File: src/main/java/org/opensearch/forecast/model/ForecastTask.java [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+        this.workerNode = input.readOptionalString();
+        if (input.readBoolean()) {
+            this.user = new User(input);
+        } else {
+            user = null;
+        }
+        // Below are new fields added since AD 1.1
+        if (input.available() > 0) {
+            if (input.readBoolean()) {
+                this.dateRange = new DateRange(input);
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/forecast/model/ForecastTask.java:73]: For something like `dateRange` are you keeping it here for bwc with the shared config and job indices?

File: src/main/java/org/opensearch/ad/model/ADTask.java [hunk 6]
```diff
@@ -613,185 +368,40 @@ public static ADTask parse(XContentParser parser, String taskId) throws IOExcept
...
 
-    public void setLatest(Boolean latest) {
-        isLatest = latest;
-    }
-
-    public void setLastUpdateTime(Instant lastUpdateTime) {
-        this.lastUpdateTime = lastUpdateTime;
+    @Override
+    public String getEntityModelId() {
+        return entity == null ? null : entity.getModelId(getConfigId()).orElse(null);
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/ad/model/ADTask.java:405]: Why is there a `getConfigID` here but in forecasting its just `.getModelId(configID)`

File: src/main/java/org/opensearch/forecast/model/ForecastTask.java [hunk 7]
```diff
@@ -0,0 +1,403 @@
+/*
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/forecast/model/ForecastTask.java:1]: nit: can use shorter license on new file
```
/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
```

**Comment by amitgalitz** [comment]
> > Have you done any testing with frontend for this branch?
> 
> I haven't. The frontend main branch CI will test backend main branch. If there is anything broken, I'll fix it. Also, since we won't release this branch until forecasting frontend is finished, I think we can leave the testing later.

got it, was mostly referring to the comment regarding this file in the original issue: https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/710e3c4ddadf3689b7f014b5df88f2a82b8db999/public/pages/DetectorDetail/utils/constants.ts

File: src/main/java/org/opensearch/timeseries/model/TimeSeriesTask.java [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+
+    public boolean isHistoricalTask() {
+        return taskType.startsWith(TimeSeriesTask.HISTORICAL_TASK_PREFIX);
+    }
+
+    /**
+     * Get detector level task id. If a task has no parent task, the task is detector level task.
+     * @return detector level task id
+     */
+    public String getDetectorLevelTaskId() {
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/timeseries/model/TimeSeriesTask.java:207]: just confirming this is to check between forecaster and detector task?

File: src/main/java/org/opensearch/forecast/model/ForecastTask.java (hunk 7 above)
Comment (kaituo) [reply src/main/java/org/opensearch/forecast/model/ForecastTask.java:1]: Changed

File: src/main/java/org/opensearch/ad/model/ADTask.java (hunk 6 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/model/ADTask.java:405]: I am gonna replace getConfigId() with configID to make it consistent.

Comment (amitgalitz) [reply src/main/java/org/opensearch/ad/model/ADTask.java:405]: sweet, and then it can be refactored to TimeSeriesTask I assume

**Comment by kaituo** [comment]
> > > Have you done any testing with frontend for this branch?
> > 
> > 
> > I haven't. The frontend main branch CI will test backend main branch. If there is anything broken, I'll fix it. Also, since we won't release this branch until forecasting frontend is finished, I think we can leave the testing later.
> 
> got it, was mostly referring to the comment regarding this file in the original issue: https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/710e3c4ddadf3689b7f014b5df88f2a82b8db999/public/pages/DetectorDetail/utils/constants.ts

created an issue https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/issues/581 and will test later.

File: src/main/java/org/opensearch/timeseries/model/TimeSeriesTask.java (hunk 8 above)
Comment (kaituo) [reply src/main/java/org/opensearch/timeseries/model/TimeSeriesTask.java:207]: Good catch. Renamed to getConfigLevelTaskId

File: src/main/java/org/opensearch/ad/model/ADTask.java (hunk 6 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/model/ADTask.java:405]: yes

**Comment by kaituo** [comment]
@sudiptoguha @amitgalitz new commit made the following replacement to address compiler errors due to https://github.com/opensearch-project/OpenSearch/pull/9073

org.opensearch.common.transport.TransportAddress => org.opensearch.core.common.transport.TransportAddress
org.opensearch.transport.TransportResponse => org.opensearch.core.transport.TransportResponse
org.opensearch.common.transport.BoundTransportAddress => org.opensearch.core.common.transport.BoundTransportAddress

**Review by amitgalitz — APPROVED** [review]

**Pull request merged**